### PR TITLE
Allow ignoring tenants

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Part of https://github.com/neondatabase/neon/issues/2537

The feature 
* adds another management command to do detach without removing local tenant files (`ignore`)
* modifies attach operation to pick up such files, if asked (requires an extra key for that, so old behavior is preserved)

It looks like there's no sense to add a separate command for reattaching, since it will be doing the same essentially.

The PR is a draft, since it does not work well in cases, when some layers are missing locally and we try to start the timeline from local files and download from the remote storage.
I have a test for this and working on a fix, but wonder if that's really needed at this stage? The "on demand" PR is huge and I try to keep this PR minimal to lessen the conflicts.
Yet, after discussing it with @problame before, I've got an idea that we need some version of this feature for the retry-less attach that comes with the first version of the refactoring, and the basic functionality works (as another test added here shows).

Should we merge it as is, faster? Or should we fix it properly and have a bigger diff?

### CODE REVIEW

File: pageserver/src/http/routes.rs [hunk 1]
```diff
@@ -355,112 +355,100 @@ async fn get_lsn_by_timestamp_handler(request: Request<Body>) -> Result<Response
 async fn tenant_attach_handler(request: Request<Body>) -> Result<Response<Body>, ApiError> {
     let tenant_id: TenantId = parse_request_param(&request, "tenant_id")?;
     check_permission(&request, Some(tenant_id))?;
+    let allow_local_files = query_param_present(&request, "allow-local-files");
 
-    info!("Handling tenant attach {tenant_id}");
+    info!("Handling tenant attach {tenant_id}, local files allowed: {allow_local_files}");
 
-    tokio::task::spawn_blocking(move || match tenant_mgr::get_tenant(tenant_id, false) {
+    let mut attaching_empty_tenant = false;
```
Comment (SomeoneToIgnore) [inline pageserver/src/http/routes.rs:362]: I start to dislike this special case more and more; looks like the only "valid" timeline-less state of a tenant is when we create one (it gets created without timelines).
Otherwise, console operations never allow us to have a tenant without timelines, so should we remove such tenants on startup and never allow in the pageserver?

Might bring the "initial timeline" concept back, we used to create both tenant and a timeline before removed this.

**Review by hlinnaka — COMMENTED** [review]
I didn't look closely at the details, as PR #2785 will change the attach code significantly so I think this will need heavy changes to rebase on top of that. But the general idea and endpoint API make sense to me.

**Comment by problame** [comment]
> The PR is a draft, since it does not work well in cases, when some layers are missing locally and we try to start the timeline from local files and download from the remote storage.

The `allow-local-files=true` seems pretty unsafe to me.
Why is that?
The pageserver has no means to tell whether the tenant's local filesystem state is the same or ahead of the remote state.
Remote state here means _all_ timelines of the tenant.
We cannot allow the pageserver to operate on a tenant if we aren't sure about this.
For example, if the remote has a timeline that the local filesystem state doesn't have, for whatever reason, then local GC will not respect that timeline's branch point LSN, which will eventually lead to data loss.

> Yet, after discussing it with @problame before, I've got an idea that we need some version of this feature for the retry-less attach that comes with the first version of the refactoring, and the basic functionality works (as another test added here shows).

Yes, #2785  will land without download retries, so, if there's a download failure during tenant attach, the tenant goes into Broken state.
#2785 adds crash-safe resumable tenant attach, though.
All we need is a way to perform the following operations:
1. drop the Broken tenant's in-memory representation
2. run #2785 's equivalent of the `collect_timelines_for_tenant` that you added in this PR
Step (2) will observe the attach marker file and resume the attach operation.

---

In talked to @LizardWizzard earlier today. Looks like he'll be the first person to exercise the new retry-less attach path, as part of the AWS account migration effort. 
He'll make the call on how urgent which aspects of this PR are.

As outlined above, my perspective is that we could drop the `allow-local-files`, once #2785 lands.

**Comment by hlinnaka** [comment]
> > 
> 
> The `allow-local-files=true` seems pretty unsafe to me. Why is that? The pageserver has no means to tell whether the tenant's local filesystem state is the same or ahead of the remote state. Remote state here means _all_ timelines of the tenant.

I didn't look closely how this actually works, but here's my mental model on how we could/should make use of existing local files:

1 Start from a clean slate by performing a "normal" attach with no local files. (rename away the old tenant-directory, for example, so that it doesn't confuse us)
2. After that's finished and we know what remote files we need to download, for each such file, check if it exists in the old local directory and copy/move it from there instead.

So in a nutshell, treat the old local directory as just a local cache to skip the downloading for those files.

Once we have on-demand download, we won't need to download anything at attach, but the same principle still works.
